mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml

```diff
@@ -1336,7 +1336,7 @@ structured_op: !LinalgStructuredOpConfig
     name: C
     kind: output_tensor
     type_var: U
-    shape_map: affine_map<()[s0, s1, s2] -> (s2, s1)>
+    shape_map: affine_map<()[s0, s1, s2] -> (s1, s2)>
```

Comment:
Why would `K` be in the middle?

Reply:
I'm realising that `shape_map` != `indexing_maps` and that all my thinking was wrong. Sorry about the confusion.

What's the intended use of `shape_map`? Is that required specifically for https://mlir.llvm.org/docs/Dialects/Linalg/OpDSL/?

Reply:
Thanks!

So this PR doesn't change the `indexing_maps` - these seem to be actually correct. And `shape_map` seem to be dropped when converting to tablegen. I guess this is only fixing OpDSL?

@JerryShih How can we test/check/verify this change?

Reply:
I don't think we have the consistent meaning for s0, s1 and s2. They are just the symbol names.
This `LinalgNamedStructuredOps.yaml` file is generated from `mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py`.

Here is the document:
https://mlir.llvm.org/docs/Dialects/Linalg/OpDSL/

In this pr, I update the `matmul_transpose_a` def in `core_named_ops.py` to follow the same dim order as `batch_matmul_transpose_a`.

https://github.com/llvm/llvm-project/blob/5fc7342a836cea6409719bdbdb5b69e8e4e7c570/mlir/python/mlir/dialects/linalg/opdsl/ops/core_named_ops.py#L530-L532

Reply:
> So this PR doesn't change the `indexing_maps` - these seem to be actually correct. And `shape_map` seem to be dropped when converting to tablegen. I guess this is only fixing OpDSL?
> 

Yes, the input/output's affine_map doesn't used in mlir-linalg-ods-gen tool now. I just try to fix the weird declarion in `core_named_ops.py`.

> @JerryShih How can we test/check/verify this change?
Sorry, I don't know how to test/verify this update. It's just try to fix the matrix dim name.

Reply:
@ftynse 
> This can be tested by manually writing an instance of `linalg.matmul_tranpose_a` taking tensors of certain fixed size (says M=10,N=20,K=30) and making sure it passes the verifier. Without this change, it wouldn't because the dimension sizes wouldn't match.

There is already a `linalg.matmul_tranpose_a` test inside.
https://github.com/llvm/llvm-project/blob/ed4e75d5e5ada30c37c57df032378a77e6dd598e/mlir/test/Dialect/Linalg/named-ops.mlir#L1193-L1200

And this pr is "nfc" patch. It doesn't change the correctness of `linalg.matmul_tranpose_a` op.

Here is the affine_map used in yaml op generator.
https://github.com/llvm/llvm-project/blob/ed4e75d5e5ada30c37c57df032378a77e6dd598e/mlir/tools/mlir-linalg-ods-gen/mlir-linalg-ods-yaml-gen.cpp#L834

It looks like only uses the affine_map in `indexing_maps` 

https://github.com/llvm/llvm-project/blob/5fc7342a836cea6409719bdbdb5b69e8e4e7c570/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml#L1099-L1103

Use linalg.matmul as the example:
```
vim ./llvm-project/tools/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yamlgen.cpp.inc

....
ArrayAttr MatmulOp::getIndexingMaps() {
  static const char memoizeAttr[] = "linalg.memoized_indexing_maps";
  ArrayAttr cached = getOperation()->getAttrOfType<ArrayAttr>(memoizeAttr);
  if (cached)
    return cached;

  MLIRContext *context = getContext();
  auto symbolBindings = getSymbolBindings(*this);
  SmallVector<AffineMap> maps;
  maps.push_back(llvm::cast<AffineMapAttr>(mlir::parseAttribute("affine_map<(d0, d1, d2)[s0, s1, s2] -> (d0, d2)>", context)).getValue());
  maps.back() = simplifyAffineMap(maps.back().replaceDimsAndSymbols({}, symbolBindings, 3, 0));
  maps.push_back(llvm::cast<AffineMapAttr>(mlir::parseAttribute("affine_map<(d0, d1, d2)[s0, s1, s2] -> (d2, d1)>", context)).getValue());
  maps.back() = simplifyAffineMap(maps.back().replaceDimsAndSymbols({}, symbolBindings, 3, 0));
  maps.push_back(llvm::cast<AffineMapAttr>(mlir::parseAttribute("affine_map<(d0, d1, d2)[s0, s1, s2] -> (d0, d1)>", context)).getValue());
  maps.back() = simplifyAffineMap(maps.back().replaceDimsAndSymbols({}, symbolBindings, 3, 0));
  cached = Builder(context).getAffineMapArrayAttr(maps);
  getOperation()->setAttr(memoizeAttr, cached);
  return cached;
}
```
I don't see the op def which is related the affine_map in `args` parts.

This pr only update the `args` parts which is not used in yaml generator.
https://github.com/llvm/llvm-project/blob/5fc7342a836cea6409719bdbdb5b69e8e4e7c570/mlir/include/mlir/Dialect/Linalg/IR/LinalgNamedStructuredOps.yaml#L1079-L1098

Reply:
> This pr only update the args parts which is not used in yaml generator.

Where is it used then? We need to find that place and add a test there. If there is no such place, we should drop this altogether.

Reply:
Note, we (Intel Labs) are working on a rewrite of matmul in table-gen, so that these variants are not relevant any more. They should continue here in case there's usage downstream, but hopefully they'll become unused soon, so we can remove them.

The new matmul will have the proposed optional affine map to encode transpose and broadcast on the `matmul` operations themselves. Anyone using them as before will not be impacted, anyone using `*_matmul_transposed_*` can start using the new `*_matmul { #transpose_map }` variants.

@shahidact